misoc/cores/icap.py

```diff
@@ -0,0 +1,152 @@
+
+        # rising edge detection
+        edge = Signal(2)
+        self.sync += [
+            edge[1].eq(edge[0]),
+            edge[0].eq(self.iprog.re)
+        ]
+
+        fsm = FSM()
+        self.submodules += fsm
+        fsm.act(0,
+            NextValue(rdwrb, 1),
+            NextValue(csib, 1),
+            If(edge == 0b01,
+                NextState(1)
+            )
+        )
+        fsm.act(1,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 1),
+            NextState(2)
+        )
+        fsm.act(2,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[0]),
+            NextState(3)
+        )
+        fsm.act(3,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[1]),
+            NextState(4)
+        )
+        fsm.act(4,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[2]),
+            NextState(5)
+        )
+        fsm.act(5,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[3]),
+            NextState(6)
+        )
+        fsm.act(6,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[4]),
+            NextState(7)
+        )
+        fsm.act(7,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[5]),
+            NextState(8)
+        )
+        fsm.act(8,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[6]),
+            NextState(9)
+        )
+        fsm.act(9,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[7]),
+            NextState(10)
+        )
+        fsm.act(10,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[8]),
+            NextState(11)
+        )
+        fsm.act(11,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[9]),
+            NextState(12)
+        )
+        fsm.act(12,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextValue(i, iprog_command_seq_i[10]),
+            NextState(13)
+        )
+        fsm.act(13,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextState(14)
+        )
+        fsm.act(14,
+            NextValue(rdwrb, 0),
+            NextValue(csib, 0),
+            NextState(0)
+        )
+
+        if platform in {'kasli', 'phaser'}:
+            # ICAPE2 primitive module
+            self.specials += Instance("ICAPE2", name="ICAPE2_inst",
+                p_ICAP_WIDTH = ICAP_WIDTH,
+
+                i_CLK = ClockSignal(),
```

Comment:
Won't this violate the maximum clock frequency specification for the ICAP when the system clock is 125MHz, which it usually is?

Reply:
I suggest using BUFR to divide the clock and create a slower domain, and using ``PulseSynchronizer`` to transfer the ``iprog.re`` pulse into the slower clock domain.

Reply:
I tested and it works on both Kasli and Metlino.

Reply:
That's not good enough reason, it may not work reliably on all boards across the temperature range; the datasheet spec is there for a reason.

Reply:
After added `PulseSynchronizer`, in the generated Verilog file, I saw that `icap_state <= icap_next_state` is still in the clock domain of `sys_clk`. I think this is not right but not sure how to correct it.

Reply:
Use ClockDomainsRenamer on the FSM